File: src/claude.ts
import { spawn } from 'node:child_process';
import { execa } from 'execa';

async function resolveClaudeBin(): Promise<string> {
  try {
    const { stdout } = await execa('which', ['claude']);
    return stdout.trim();
  } catch {
    throw new Error(
      'Could not find `claude` on PATH. Install Claude Code: https://docs.claude.com/en/docs/claude-code',
    );
  }
}

export async function launchClaude(cwd: string, extraArgs: string[] = []): Promise<number> {
  const bin = await resolveClaudeBin();
  return await new Promise<number>((resolve, reject) => {
    const child = spawn(bin, extraArgs, {
      cwd,
      stdio: 'inherit',
      env: process.env,
    });
    child.on('error', reject);
    child.on('exit', (code) => resolve(code ?? 0));
  });
}

File: zig/src/claude.zig
//! Launching Claude Code in a worktree.

const std = @import("std");
const Io = std.Io;
const exec = @import("exec.zig");

pub const Error = error{ClaudeNotFound} || std.mem.Allocator.Error;

fn resolveBin(gpa: std.mem.Allocator, io: Io) Error![]u8 {
    return exec.capture(gpa, io, &.{ "which", "claude" }, null) catch
        return Error.ClaudeNotFound;
}

/// Hands the terminal to Claude Code and returns its exit status.
pub fn launch(
    gpa: std.mem.Allocator,
    io: Io,
    cwd: []const u8,
    extra_args: []const []const u8,
) !u8 {
    const bin = try resolveBin(gpa, io);

    var argv: std.ArrayList([]const u8) = .empty;
    try argv.append(gpa, bin);
    try argv.appendSlice(gpa, extra_args);

    return exec.inherit(io, argv.items, cwd);
}
